cou/apps/app.py

```diff
@@ -182,7 +305,251 @@ def _get_os_origin(self) -> str:
         """
         for origin in ("openstack-origin", "source"):
             if self.config.get(origin):
+                self.origin_setting = origin
                 return self.config[origin].get("value", "")
 
         logger.warning("Failed to get origin for %s, no origin config found", self.name)
         return ""
+
+    async def _check_upgrade(self, target: OpenStackRelease) -> None:
```

Comment:
failing to complete the upgrade is pretty serious; the overall cloud upgrade should probably halt. Though this will be the responsibility of the execute phase, should we think of throwing an exception here?

Reply:
I thought about that too. I've changed to raise `ApplicationError`